README
TeX2page is a Scheme or Common Lisp program for making Web pages
from TeX documents.  

Please see http://ds26gte.github.io/tex2page/index.html for what
it can do.

Installation instructions are at
http://github.com/ds26gte/tex2page/blob/master/INSTALL.


history
Last change 2022-07-04

2022 Jul 4

Ported to Clasp.

2020 Nov 15

For LaTeX, emitting to HTML shd be turned off between \documentclass &
  \begin{document}, not from the very start!
Have \usepackage find & load its arg, but only if it's in cwd or
  TEX2PAGEINPUTS.
find-tex-file gets optarg to skip kpathsea.
  It also has TEX2PAGEINPUTS only shadow kpathsea, not displace it completely.

2020 Nov 14

bibtex & makeindex refuse to work if arg is given as absolute pathname.
  Causes problem if .tex2page.hdir has abs pathname. Work around it.
Added LaTeX Companion 2/e to bibliog.

Def of \epsfbox in luatex needs changed: 
  check for \directlua not \pdfoutput;
  \input miniltx, graphicx, not supp-pdf.

tex2page.lisp: Rewrite scm-output-verbaitm-comment's loop so Racket translation is correct.
The if-stack entries have three values now: 
- #t if currently in the true branch of a running \if;
- #f if currently in the false branch of a running \if;
- 'skip if currently in a non-running \if (i.e., in the false branch of wider \if).
This avoids problems if even the minimal parsing of a skippable nested \if could cause error.
tgtermes.tex: Don't set math fonts -- the encoding doesn't match (e.g., \Gamma fails).
Don't make GitHub pages from docs/ directory; keeps repo sparer.

2020 May 06

Use macros {string-append, list->string, string->list,
string-length} in tex2page.lisp instead of {concatenate, length}.
Makes for less clumsy translation to Racket.

2020 Apr 14

tex2page.rkt: Simplify #! line: avoid -m; use (current-command-line-arguments)
in body of code instead.

2020 Mar 29

If LaTeX format can be determined (which it will before second run), do not render preamble 
(stuff before \begin{document}).

2020 Jan 31

tex2page.lisp: Replace case with tags t/nil to cond to avoid problem translating to Racket.
do-cr can have *tabular-stack* when \\ is used outside an environment (as in \title).

2019 Nov 25

Cleaner def for system-with-visual so SBCL doesn't show unnecessary diagnostic.

./configure --dialect=sbcl should at least make $LISP unnecessary.

2019 Nov 22

Accept fancyvrb's \Verbatim[*], \DefineShortVerb.

Move loop return in do-verb-braced to tail-call position to prevent error in Racket translation.

2019 Jan 26

Ignore \set{main,sans,mono}font with args.

2018 Jul 3

Make index.tex processable by LuaTeX without recourse to external
MetaPost.

2018 Jun 29

Added \mplibcode ... \endmplibcode.

Default TeX program used by TeX2page (for making images) is
LuaTeX.

2018 Jun 28

color.sty won't define 'hsb' model for LuaTeX, which is a basis
for defining several new models in coloraug.tex. Added two
subsidiary files for coloraug.tex: colorlua.tex for LuaTeX, and
colorxe.tex for others (eg, XeTeX).

Move Lua portion of coloraug.tex ot colormodels.tex.

Allow index.tex to be processable by LuaTeX.

2018 Jun 14

Don't set _ catcode to 13 as it messes with loading of some
standard files. Instead let user optionally call
\allowunderscoreintext.

tgtermes.tex: font features separated by semicolon, not colon.
(embolden=1 doesn't seem to work.)

2017 Dec 22

Footnote counter shouldn't be 0'd for each chapter in single-page
mode.

2017 Dec 14

Properly qualify <img src=...> (aux file dir could be set).

2017 Dec 13

Put generated HTML doc in docs/ subdirectory.

2017 Dec 4

\bordermatrix = \pmatrix, for now.

Ignore \chapter etc.'s optional arg for now.

\hbox, \mbox can be followed by a single character token, i.e., not necessarily
a group.

2017 Dec 3

Accept \left. and \right. as null delimiters.

When finding \parindent (do-indent, do-plain-item, do-textindent), use
the-dimen, not find-dimen!

When \TZPsinglepage: (1) Embed basic CSS style within document instead of
generating CSS file. (2) Embed images as data URIs instead of pointing to
files. This allows for single-page HTMLs to be self-sufficient when emailing to
a Kindle.

do-end-equation: Make sure correct environment name is identified and matched.

$$: Allow for single $ to occur even when not nested instead braces. This is
because {tabular} allows this situation.

2017 Feb 7

Setting \TZPsinglepage and \TZPslides shouldn't be persistent in the aux file.

2017 Jan 29

Keep \write18's arg in one line to satisfy ECL's 'system'.

2017 Jan 28

Added color models 'hsl', 'Hsl', even though standard LaTeX packages don't have
them.

2017 Jan 27

'system' won't work to open visual editor in ECL, MKCL; use run-program
instead.

2017 Jan 26

Convert HSB, gray, and wave colors to HSL rather than RGB.

2017 Jan 25

Rewrote count/dimen/box/toks/stream code to use numbered registers (more
TeX-like). Defining names for these (\countdef, etc) allocates a new register.

2017 Jan 23

Added \lower, \raise. These expect a box, and signal error if box not found.

2017 Jan 22

do-para: Use output-port's spacing buffer to avoid most of the empty <p></p>'s.

2017 Jan 21

Add \wlog, which writes only to log file, not to console.

Like TeX, write current time on startup only to log file, not to console.

When outputting to HTML, use a buffered stream. Buffer stores only spaces,
which user can nix with \unskip.

2017 Jan 20

tex-def-pat-prim added to facilitate adding standard defs with parameters.

\loop, \iterate, \repeat added.

do-let inside false world should take care that both lhs and rhs are skipped
completely.

get-def-parameters (better name than get-def-arguments) is now a properly
tail-recursive loop.

Since get-ctl-seq doesn't read newline, it's now on primitive control sequences
to read past such newlines if necessary.

\colorbox takes optional model argument.

Use CSS for <sup> as otherwise numbered footnotes break leading. GIMP should be
all-uppercase.

2017 Jan 19

Consecutive \space, '\ ' shouldn't collapse: Use U+200B (zero-width space) to
this end.

Clean up ignorespaces: Four scenarios:
1. Eat space upto first newline (exclusive).
2. Eat upto first newline (inclusive).
3. Eat upto second newline (exclusive): indeed double it if second

colormodels.lua
-- Last modified 2018-06-28
-- Dorai Sitaram

local function hsl360_to_rgb(H, s, L)
  local c = (1 - math.abs(2*L - 1)) * s
  local x = c * (1 - math.abs(((H/60) % 2) - 1))
  local m = L - c/2
  local r1, g1, b1 = 0, 0, 0
  if H < 60 then r1 = c; g1 = x
  elseif H < 120 then r1 = x; g1 = c
  elseif H < 180 then g1 = c; b1 = x
  elseif H < 240 then g1 = x; b1 = c
  elseif H < 300 then r1 = x; b1 = c
  else r1 = c; b1 = x
  end
  return "rgb", { r1+m, g1+m, b1+m }
end

local function hsl_to_rgb(h, s, L)
  return hsl360_to_rgb(h * 360, s, L)
end

local function hsb360_to_rgb(h, s, b)
  local L = .5 * b * (2 - s)
  local s2 =  (b * s) / (1 - math.abs(2*L - 1))
  return hsl360_to_rgb(h, s2, L)
end

local function hsb_to_rgb(h, s, b)
  return hsb360_to_rgb(h*360, s, b)
end

local function hsb240_to_rgb(h, s, b)
  return hsb_to_rgb(h/240, s/240, b/240)
end

local function cmy_to_rgb(c, m, y)
  return "rgb", { 1-c, 1-m, 1-y }
end

local function html_to_rgb(xn)
  local y = xn
  local r = math.floor(y / 0x10000)
  y = y % 0x10000
  local g = math.floor(y / 0x100)
  local b = y % 0x100
  return "rgb", { r/255, g/255, b/255 }
end

local function gray15_to_gray(g)
  return "gray", { g/15 }
end

local function wavelength_to_rgb(w)
  local h
  if w <= 362.857 then h = 5
  elseif w < 440 then h = 4 + (w - 440)/-60
  elseif w < 490 then h = 4 - (w - 440)/50
  elseif w < 510 then h = 2 + (w - 510)/-20
  elseif w < 580 then h = 2 - (w - 510)/70
  elseif w < 645 then h = (w - 645)/-65
  else h = 0
  end
  h = h * 1/6
  local b = 0
  if w <= 362.857 then b = 0
  elseif w < 420 then b = .3 + .7 * (w - 380)/40
  elseif w <= 700 then b = 1
  elseif w < 814.285 then b = .3 + .7 * (w - 780)/-80
  else b = 0
  end
  return hsb360_to_rgb(h*360, 1, b)
end

function convert_to_known_color_model(model, a)
  if model == 'hsb' then
    return hsb_to_rgb(a[1], a[2], a[3])
  elseif model == 'Hsb' then
    return hsb360_to_rgb(a[1], a[2], a[3])
  elseif model == 'HSB' then
    return hsb240_to_rgb(a[1], a[2], a[3])
  elseif model == 'hsl' then
    return hsl_to_rgb(a[1], a[2], a[3])
  elseif model == 'Hsl' then
    return hsl360_to_rgb(a[1], a[2], a[3])
  elseif model == 'cmy' then
    return cmy_to_rgb(a[1], a[2], a[3])
  elseif model == 'HTML' then
    return html_to_rgb(a[1])
  elseif model == 'Gray' then
    return gray15_to_gray(a[1])
  elseif model == 'wave' then
    return wavelength_to_rgb(a[1])
  else
    return model, a
  end
end

function convert_to_csv(a)
  local s = '{'
  for i = 1, #a do
    if i > 1 then
      s = s .. ', '
    end
    s = s .. string.format("%.4f", a[i])
  end
  s = s .. '}'
  return s
end 
